src/persist-client/src/impl/machine.rs

```diff
@@ -758,15 +761,86 @@ mod tests {
 
             let state = Arc::new(Mutex::new(DatadrivenState::default()));
             let cpu_heavy_runtime = Arc::new(CpuHeavyRuntime::new());
+            let write = Arc::new(Mutex::new(
+                client
+                    .open_writer::<String, (), u64, i64>(shard_id)
+                    .await
+                    .expect("invalid shard types"),
+            ));
 
             f.run_async(move |tc| {
                 let shard_id = shard_id.clone();
                 let client = client.clone();
                 let state = Arc::clone(&state);
                 let cpu_heavy_runtime = Arc::clone(&cpu_heavy_runtime);
+                let write = Arc::clone(&write);
                 async move {
                     let mut state = state.lock().await;
+
                     match tc.directive.as_str() {
+                        // Returns whether a blob contains a given batch. The batch
+                        // must contain only a single key.
+                        "blob-batch-exists" => {
+                            let input = get_arg(&tc.args, "input").expect("missing input");
+                            let batch = state.batches.get(input).expect("unknown batch").clone();
+                            let key = batch.keys.get(0).expect("unknown batch part");
+
+                            assert_eq!(batch.keys.len(), 1);
+
+                            let blob = client
+                                .blob
+                                .get(&key.complete(&shard_id))
+                                .await
+                                .expect("blob read failed");
+
+                            format!("{}\n", blob.is_some())
```

Comment:
in fact, now that I think about it, can we avoid "blob-batch-exists" by modifying "fetch-batch" to print something like "missing" if a key doesn't exist

Reply:
oh! I thought it was just being pulled from DatadrivenState and not also calling into `Blob::get`. we can definitely just have `fetch-batch`